io/js/src/main/scala/fs2/io/compressionplatform.scala

```diff
@@ -59,25 +62,98 @@ private[io] trait compressionplatform {
+      ): Stream[F, Byte] => Stream[F, (Stream[F, Byte], Deferred[F, Chunk[Byte]])] = in => {
         val options = zlibMod
           .ZlibOptions()
           .setChunkSize(inflateParams.bufferSizeOrMinimum.toDouble)
+
         Stream
-          .resource(suspendReadableAndRead() {
-            (inflateParams.header match {
-              case ZLibParams.Header.GZIP => zlibMod.createGunzip(options)
-              case ZLibParams.Header.ZLIB => zlibMod.createInflate(options)
-            }).asInstanceOf[Duplex]
-          })
+          .resource {
+            suspendReadableAndRead() {
+              (inflateParams.header match {
+                case ZLibParams.Header.GZIP => zlibMod.createInflateRaw(options)
+                case ZLibParams.Header.ZLIB => zlibMod.createInflate(options)
+              }).asInstanceOf[Duplex]
+            }
+          }
           .flatMap { case (inflate, out) =>
-            out
-              .concurrently(in.through(writeWritable[F](inflate.pure.widen)))
-              .onFinalize(
-                F.async_[Unit](cb => inflate.asInstanceOf[zlibMod.Zlib].close(() => cb(Right(()))))
-              )
+            Stream.resource(SuspendedStream(in)).flatMap { suspendedIn =>
+              Stream.eval(F.ref(Chunk.empty[Byte])).flatMap { lastChunk =>
+                Stream.eval(F.ref(0)).flatMap { bytesPiped =>
+                  Stream.eval(F.deferred[Int]).flatMap { bytesWritten =>
+                    Stream.eval(F.deferred[Chunk[Byte]]).flatMap { trailerChunk =>
+                      val trackedStream =
+                        suspendedIn.stream.chunks.evalTap { chunk =>
+                          bytesPiped.update(_ + chunk.size) >> lastChunk.set(chunk)
+                        }.unchunks
+                      val inflated = out
+                        .concurrently(
+                          trackedStream
+                            .through(writeWritable[F](inflate.asInstanceOf[Writable].pure))
+                        )
+                        .onFinalize {
+                          F.delay(inflate.asInstanceOf[zlibMod.Zlib].bytesWritten.toLong.toInt)
+                            .flatMap(bytesWritten.complete) >>
+                            F.async_[Unit] { cb =>
+                              inflate.asInstanceOf[zlibMod.Zlib].close(() => cb(Right(())))
+                            }
+                        }
+
+                      Stream
+                        .resource {
+                          F.background {
+                            (bytesWritten.get, bytesPiped.get, lastChunk.get).mapN {
+                              (bytesWritten, bytesPiped, lastChunk) =>
+                                val bytesAvailable = bytesPiped - bytesWritten
+                                val headTrailerBytes = lastChunk.takeRight(bytesAvailable)
+                                val bytesToPull = trailerSize - headTrailerBytes.size
+                                if (bytesToPull > 0) {
```

Comment:
I *think* I got the math right, here. But I'm not sure if this case happens during the tests.